File: frontend/src/shared/api/protocols.ts
import { axiosInstance } from '../config/axios';
import type { PaginatedResponse } from './research';
import type { Sample } from './samples';

export interface Protocol {
  id: number;
  test_protocol_number?: string;
  test_protocol_date?: string;
  is_accredited: boolean;
  sampling_act_number: string;
  issued?: string;
  approved?: string;
  issued_position?: string;
  approved_position?: string;
  protocol_template_id?: number;
  laboratory_id: number;
  department_id?: number;
  samples?: number[];
  samples_data?: Sample[];
  laboratory_name?: string;
  department_name?: string;
  formatted_protocol_number?: string;
  has_undeleted_calculations?: boolean;
  created_at: string;
  updated_at: string;
  deleted_at?: string;
}

export interface ProtocolCreate {
  test_protocol_number?: string;
  test_protocol_date?: string;
  is_accredited?: boolean;
  sampling_act_number: string;
  issued?: string;
  approved?: string;
  issued_position?: string;
  approved_position?: string;
  protocol_template_id?: number;
  laboratory_id: number;
  department_id?: number;
  samples?: number[];
}

export interface ProtocolUpdate {
  test_protocol_number?: string;
  test_protocol_date?: string;
  is_accredited?: boolean;
  sampling_act_number?: string;
  issued?: string;
  approved?: string;
  issued_position?: string;
  approved_position?: string;
  protocol_template_id?: number;
  laboratory_id?: number;
  department_id?: number;
  samples?: number[];
}

export interface ProtocolFilters {
  test_protocol_number?: string;
  sampling_act_number?: string;
  is_accredited?: boolean;
  test_protocol_date_from?: string;
  test_protocol_date_to?: string;
  test_protocol_date_search?: string;
  created_at_from?: string;
  created_at_to?: string;
  [key: string]: unknown;
}

export interface ProtocolTemplate {
  id: number;
  name: string;
  file_name: string;
  version: string;
  laboratory_id: number;
  department_id?: number;
  laboratory_name?: string;
  department_name?: string;
  created_at: string;
  updated_at: string;
  deleted_at?: string;
}

export interface CellStyle {
  fontWeight?: string;
  fontStyle?: string;
  fontSize?: string;
  textAlign?: 'left' | 'center' | 'right';
}

export const protocolsApi = {
  getProtocols: async (
    page?: number,
    pageSize?: number,
    filters?: ProtocolFilters,
    sorting?: { sort_by?: string; sort_order?: 'asc' | 'desc' },
    laboratoryId?: number,
    departmentId?: number
  ): Promise<PaginatedResponse<Protocol>> => {
    const params: Record<string, unknown> = {};

    if (page !== undefined) {
      params.page = page;
    }

    if (pageSize !== undefined) {
      params.page_size = pageSize;
    }

    if (laboratoryId) {
      params.laboratory_id = laboratoryId;
    }

    if (departmentId) {
      params.department_id = departmentId;
    }

    if (filters?.test_protocol_number) {
      params.search = filters.test_protocol_number;
    }

    if (filters?.sampling_act_number) {
      params.search_sampling_act = filters.sampling_act_number;
    }

    if (filters?.search_samples) {
      params.search_samples = filters.search_samples;
    }

    if (filters?.is_accredited !== undefined && filters?.is_accredited !== null) {
      params.is_accredited = filters.is_accredited;
    }

    if (filters?.test_protocol_date_from) {
      params.test_protocol_date_from = filters.test_protocol_date_from;
    }
    if (filters?.test_protocol_date_to) {
      params.test_protocol_date_to = filters.test_protocol_date_to;
    }
    if (filters?.test_protocol_date_search) {
      params.search_date = filters.test_protocol_date_search;
    }
    if (filters?.created_at_from) {
      params.created_at_from = filters.created_at_from;
    }
    if (filters?.created_at_to) {
      params.created_at_to = filters.created_at_to;
    }

    if (sorting?.sort_by) {
      params.sort_by = sorting.sort_by;
      params.sort_order = sorting.sort_order || 'desc';
    }

    const response = await axiosInstance.get<PaginatedResponse<Protocol>>('/api/protocols/', {
      params,
    });
    return response.data;
  },

  getProtocol: async (id: number): Promise<Protocol> => {
    const response = await axiosInstance.get<Protocol>(`/api/protocols/${id}/`);
    return response.data;
  },

  createProtocol: async (data: ProtocolCreate): Promise<Protocol> => {
    const response = await axiosInstance.post<Protocol>('/api/protocols/', data);
    return response.data;
  },

  updateProtocol: async (id: number, data: ProtocolUpdate): Promise<Protocol> => {
    const response = await axiosInstance.patch<Protocol>(`/api/protocols/${id}/`, data);
    return response.data;
  },

  deleteProtocol: async (id: number): Promise<void> => {
    await axiosInstance.delete(`/api/protocols/${id}/`);
  },

  generateProtocolExcel: async (protocolId: number): Promise<{ blob: Blob; filename: string }> => {
    try {
      const response = await axiosInstance.get(`/api/protocols/${protocolId}/generate-excel/`, {
        responseType: 'blob',
      });

      // Извлекаем имя файла из заголовка Content-Disposition
      let filename = `protocol_${protocolId}.xlsx`; // fallback
      // Для blob-ответов заголовки могут быть в разных форматах
      const headers = response.headers as Record<string, string> & {
        get?: (name: string) => string | null;
      };
      const contentDisposition =
        response.headers['content-disposition'] ||
        response.headers['Content-Disposition'] ||
        headers.get?.('content-disposition') ||
        null;

      if (contentDisposition) {
        // Парсим заголовок: ищем filename*=UTF-8''... или filename="..."
        const filenameMatch =
          contentDisposition.match(/filename\*=UTF-8''(.+?)(?:;|$)/i) ||
          contentDisposition.match(/filename="(.+?)"/i) ||
          contentDisposition.match(/filename=([^;]+)/i);

        if (filenameMatch) {
          try {
            // Декодируем percent-encoded имя файла
            filename = decodeURIComponent(filenameMatch[1]);
          } catch {
            filename = filenameMatch[1].replace(/['"]/g, '');
          }
        }
      }
      return { blob: response.data, filename };
    } catch (error: unknown) {
      if (error && typeof error === 'object' && 'response' in error) {
        const axiosError = error as { response?: { data?: Blob; status?: number } };
        if (
          axiosError.response?.data &&
          axiosError.response.status &&
          axiosError.response.status >= 400
        ) {
          const blob = axiosError.response.data;
          try {
            const text = await blob.text();
            const json = JSON.parse(text);
            if (json.detail) {
              throw new Error(json.detail);
            }
          } catch {
            throw error;
          }
        }
      }
      throw error;
    }
  },

  getAvailableProtocolTemplates: async (
    laboratoryId: number,
    departmentId?: number
  ): Promise<ProtocolTemplate[]> => {
    const params: Record<string, unknown> = {
      laboratory_id: laboratoryId,
    };
    if (departmentId) {
      params.department_id = departmentId;
    }
    const response = await axiosInstance.get<ProtocolTemplate[]>(
      '/api/protocol-templates/available/',
      { params }
    );
    return response.data;
  },

  getProtocolTemplates: async (
    laboratoryId?: number,
    departmentId?: number,
    page?: number,
    pageSize?: number
  ): Promise<PaginatedResponse<ProtocolTemplate>> => {
    const params: Record<string, unknown> = {};
    if (laboratoryId) {
      params.laboratory_id = laboratoryId;
    }
    if (departmentId) {
      params.department_id = departmentId;
    }
    if (page !== undefined) {
      params.page = page;
    }
    if (pageSize !== undefined) {
      params.page_size = pageSize;
    }
    const response = await axiosInstance.get<PaginatedResponse<ProtocolTemplate>>(
      '/api/protocol-templates/',
      { params }
    );
    return response.data;
  },

  getProtocolTemplate: async (id: number): Promise<ProtocolTemplate> => {
    const response = await axiosInstance.get<ProtocolTemplate>(`/api/protocol-templates/${id}/`);
    return response.data;
  },

  createProtocolTemplate: async (data: {
    name: string;
    file_name: string;
    file: string;
    laboratory_id: number;
    department_id?: number;
  }): Promise<ProtocolTemplate> => {
    const response = await axiosInstance.post<ProtocolTemplate>('/api/protocol-templates/', data);
    return response.data;
  },

  updateProtocolTemplate: async (
    id: number,
    data: Partial<ProtocolTemplate> & { file?: string }
  ): Promise<ProtocolTemplate> => {
    const response = await axiosInstance.patch<ProtocolTemplate>(
      `/api/protocol-templates/${id}/`,
      data
    );
    return response.data;
  },

  saveExcelSection: async (
    formData: FormData
  ): Promise<{ template_id?: number; error?: string }> => {
    const response = await axiosInstance.post<{ template_id?: number; error?: string }>(
      '/api/save-excel/',
      formData,
      {
        headers: {
          'Content-Type': 'multipart/form-data',
        },
      }
    );
    return response.data;
  },

  getExcelStyles: async (
    templateId: number,
    section: string
  ): Promise<{ styles?: Record<string, CellStyle>; error?: string }> => {
    const response = await axiosInstance.get<{
      styles?: Record<string, CellStyle>;
      error?: string;
    }>('/api/get-excel-styles/', {
      params: {
        template_id: templateId,
        section: section,
      },
    });
    return response.data;
  },

  getProtocolTemplateFile: async (templateId: number, section: string): Promise<ArrayBuffer> => {
    const response = await axiosInstance.get(`/api/protocol-templates/${templateId}/`, {
      params: {
        download: true,
        section: section,
      },
      responseType: 'arraybuffer',
    });
    return response.data;
  },
};

File: backend/api/protocols.py
from __future__ import annotations
from fastapi import APIRouter, Depends
from core.deps import DbSession, ProtocolListFilters, UserPermissions
from core.responses import build_attachment_response
from schemas.pagination import PaginatedResponse
from schemas.protocol import (
    ProtocolCreate,
    ProtocolResponse,
    ProtocolUpdate,
)
from services.access_control import enforce_crud_access
from services.protocol import (
    build_protocols_list_response,
    create_protocol,
    delete_protocol,
    get_protocol_detail_response,
    get_protocol_response_data,
    get_protocols,
    require_protocol_by_id,
    update_protocol,
)
from services.protocol_generator import generate_protocol_excel

router = APIRouter()


@router.get(
    "/protocols/",
    response_model=PaginatedResponse[ProtocolResponse],
    summary="Получение списка протоколов",
    description=(
        "Возвращает список протоколов с пагинацией или без. "
        "Если page и page_size не указаны, возвращает все записи. "
        "Поддерживает фильтрацию по лабораториям и подразделениям, сортировку."
    ),
    responses={200: {"description": "Список протоколов успешно получен"}},
)
# @IsAuthenticated
async def list_protocols(
    db: DbSession,
    effective: UserPermissions,
    filters: ProtocolListFilters = Depends(),
):
    """Возвращает список протоколов с пагинацией или без."""
    enforce_crud_access(
        effective,
        "protocols",
        "read",
        filters.laboratory_id,
        filters.department_id,
    )
    protocols, total, total_pages = await get_protocols(
        db,
        laboratory_id=filters.laboratory_id,
        department_id=filters.department_id,
        include_deleted=filters.include_deleted,
        page=filters.page,
        page_size=filters.page_size,
        sort_by=filters.sort_by,
        sort_order=filters.sort_order,
        is_accredited=filters.is_accredited,
        search=filters.search,
        search_date=filters.search_date,
        search_sampling_act=filters.search_sampling_act,
        search_samples=filters.search_samples,
        test_protocol_date_from=filters.test_protocol_date_from,
        test_protocol_date_to=filters.test_protocol_date_to,
        created_at_from=filters.created_at_from,
        created_at_to=filters.created_at_to,
    )

    items = await build_protocols_list_response(db, protocols)

    return PaginatedResponse(
        items=items,
        total=total,
        page=filters.page if filters.page is not None else 1,
        page_size=filters.page_size if filters.page_size is not None else total,
        total_pages=total_pages,
    )


@router.post(
    "/protocols/",
    response_model=ProtocolResponse,
    status_code=201,
    summary="Добавление нового протокола",
    description="Добавляет новый протокол на основе переданных данных.",
    responses={
        201: {"description": "Протокол успешно добавлен"},
        400: {"description": "Некорректные данные для добавления протокола"},
    },
)
# @IsAuthenticated
async def create_protocol_endpoint(
    protocol_data: ProtocolCreate,
    db: DbSession,
    effective: UserPermissions,
):
    """Добавляет новый протокол на основе переданных данных."""
    enforce_crud_access(
        effective,
        "protocols",
        "create",
        protocol_data.laboratory_id,
        protocol_data.department_id,
    )
    protocol = await create_protocol(db, protocol_data)
    return await get_protocol_response_data(db, protocol.id)


@router.get(
    "/protocols/{protocol_id}/",
    response_model=ProtocolResponse,
    summary="Получение протокола по ID",
    description="Возвращает информацию о протоколе по его идентификатору.",
    responses={
        200: {"description": "Протокол успешно получен"},
        404: {"description": "Протокол не найден"},
    },
)
# @IsAuthenticated
async def get_protocol(
    protocol_id: int,
    db: DbSession,
    effective: UserPermissions,
):
    """Возвращает информацию о протоколе по его идентификатору."""
    protocol = await require_protocol_by_id(db, protocol_id)
    enforce_crud_access(
        effective,
        "protocols",
        "read",
        protocol.laboratory_id,
        protocol.department_id,
    )
    return await get_protocol_detail_response(db, protocol_id)


@router.patch(
    "/protocols/{protocol_id}/",
    response_model=ProtocolResponse,
    summary="Обновление протокола",
    description="Обновляет существующий протокол.",
    responses={
        200: {"description": "Протокол успешно обновлен"},
        404: {"description": "Протокол не найден"},
    },
)
# @IsAuthenticated
async def update_protocol_endpoint(
    protocol_id: int,
    protocol_data: ProtocolUpdate,
    db: DbSession,
    effective: UserPermissions,
):
    """Обновляет существующий протокол."""
    protocol = await require_protocol_by_id(db, protocol_id)
    lab_id = protocol_data.laboratory_id or protocol.laboratory_id
    dept_id = protocol_data.department_id or protocol.department_id
    enforce_crud_access(effective, "protocols", "update", lab_id, dept_id)
    protocol = await update_protocol(db, protocol_id, protocol_data)
    return await get_protocol_response_data(db, protocol.id)


@router.delete(
    "/protocols/{protocol_id}/",
    status_code=204,
    summary="Удаление протокола",
    description="Выполняет мягкое удаление протокола.",
    responses={
        204: {"description": "Протокол успешно удален"},
        404: {"description": "Протокол не найден"},
    },
)
# @IsAuthenticated
async def delete_protocol_endpoint(
    protocol_id: int,
    db: DbSession,
    effective: UserPermissions,
):
    """Выполняет мягкое удаление протокола."""
    protocol = await require_protocol_by_id(db, protocol_id)
    enforce_crud_access(
        effective,
        "protocols",
        "delete",
        protocol.laboratory_id,
        protocol.department_id,
    )
    await delete_protocol(db, protocol_id)


@router.get(
    "/protocols/{protocol_id}/generate-excel/",
    summary="Генерация Excel файла протокола",
    description=(
        "Генерирует Excel-файл протокола на основе данных протокола и шаблона и возвращает его для скачивания."
    ),
    responses={
        200: {
            "description": "Excel файл успешно сгенерирован",
            "content": {"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet": {}},
        },
        404: {"description": "Протокол не найден"},
    },
)
# @IsAuthenticated
async def generate_protocol_excel_endpoint(
    protocol_id: int,
    db: DbSession,
    effective: UserPermissions,
):
    """Генерирует Excel-файл протокола на основе данных протокола и шаблона."""
    protocol = await require_protocol_by_id(db, protocol_id)
    enforce_crud_access(
        effective,
        "protocols",
        "read",
        protocol.laboratory_id,
        protocol.department_id,
    )
    content, filename = await generate_protocol_excel(db, protocol_id)
    return build_attachment_response(
        content,
        filename,
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    )
